doc/man3/EVP_PKEY_CTX_ctrl.pod

```diff
@@ -116,56 +197,182 @@ padding for RSA the algorithm identifier byte is added or checked and removed
+The EVP_PKEY_CTX_set_rsa_oaep_md() macro sets the message digest type used
+in RSA OAEP to B<md>. The padding mode must have been set to
+B<RSA_PKCS1_OAEP_PADDING>.
+
+The EVP_PKEY_CTX_get_rsa_oaep_md() macro gets the message digest type used
+in RSA OAEP to B<md>. The padding mode must have been set to
+B<RSA_PKCS1_OAEP_PADDING>.
+
+The EVP_PKEY_CTX_set0_rsa_oaep_label() macro sets the RSA OAEP label to
+B<label> and its length to B<len>. If B<label> is NULL or B<len> is 0,
+the label is cleared. The library takes ownership of the label so the
+caller should not free the original memory pointed to by B<label>.
+The padding mode must have been set to B<RSA_PKCS1_OAEP_PADDING>.
+
+The EVP_PKEY_CTX_get0_rsa_oaep_label() macro gets the RSA OAEP label to
+B<label>. The return value is the label length. The padding mode
+must have been set to B<RSA_PKCS1_OAEP_PADDING>. The resulting pointer is owned
+by the library and should not be freed by the caller.
+
+The EVP_PKEY_CTX_set_dsa_paramgen_bits() macro sets the number of bits used
 for DSA parameter generation to B<bits>. If not specified 1024 is used.
 
-The macro EVP_PKEY_CTX_set_dh_paramgen_prime_len() sets the length of the DH
+The EVP_PKEY_CTX_set_dh_paramgen_prime_len() macro sets the length of the DH
 prime parameter B<p> for DH parameter generation. If this macro is not called
-then 1024 is used.
+then 1024 is used. Only accepts lengths greater than or equal to 256.
+
+The EVP_PKEY_CTX_set_dh_paramgen_subprime_len() macro sets the length of the DH
+optional subprime parameter B<q> for DH parameter generation. The default is
+256 if the prime is at least 2048 bits long or 160 otherwise. The DH
+paramgen type must have been set to x9.42.
 
 The EVP_PKEY_CTX_set_dh_paramgen_generator() macro sets DH generator to B<gen>
 for DH parameter generation. If not specified 2 is used.
 
+The EVP_PKEY_CTX_set_dh_paramgen_type() macro sets the key type for DH
+parameter generation. Use 0 for PKCS#3 DH and 1 for X9.42 DH.
+The default is 0.
+
 The EVP_PKEY_CTX_set_dh_pad() macro sets the DH padding mode. If B<pad> is
 1 the shared secret is padded with zeroes up to the size of the DH prime B<p>.
 If B<pad> is zero (the default) then no padding is performed.
 
 EVP_PKEY_CTX_set_dh_nid() sets the DH parameters to values corresponding to
 B<nid>. The B<nid> parameter must be B<NID_ffdhe2048>, B<NID_ffdhe3072>,
 B<NID_ffdhe4096>, B<NID_ffdhe6144> or B<NID_ffdhe8192>.  This macro can be
-called during parameter or key generation.
+called during parameter or key generation. The nid must be set before
+the rfc5114 parameter.
+
+The EVP_PKEY_CTX_set_dh_rfc5114() and EVP_PKEY_CTX_set_dhx_rfc5114() are
+synonymous. When set, the RFC5114 parameters are used instead of new DH
+parameters being generated so all other DH parameters are overriden. Possible
+values are 1, 2 and 3 corresponding to RFC5114 sections 2.1, 2.2 and 2.3.
+0 can be used to clear the stored value. The nid param must have been set.
```

Comment:
The nid param must *not* have been set. Please note in this para that this is used for parameter generation and that you should create the EVP_PKEY_CTX by using `EVP_PKEY_CTX_new_id(EVP_PKEY_DHX, eng)`. Alternatively I'm wondering whether we should fix `EVP_PKEY_CTX_set_dh_rfc5114` to just use `EVP_PKEY_DH` in the public header since this is obviously a bug...but maybe people are already using it ;-(

Reply:
Since the next version is planned to be a major release, wouldn't it make sense to take the opportunity to do this kind of breaking changes?

Reply:
> Since the next version is planned to be a major release, wouldn't it make sense to take the opportunity to do this kind of breaking changes?

It will also be a goal of the next release to keep breaking changes to an absolute minimum.

> Also the EVP_PKEY_DHX flag is present both in the header and in the implementation of dh_pkey_paramgen. Maybe it not a bug but intentional behavior?

I would expect to see it in `dh_pkey_paramgen`. The generated parameters are X9.42 style parameters (which correspond to EVP_PKEY_DHX). There is a difference between the type used to create the EVP_PKEY_CTX, and the type of the parameters that are eventually generated using that EVP_PKEY_CTX.

Elsewhere when we generate X9.42 parameters (not using RFC5114), we allow you to create an EVP_PKEY_CTX using `EVP_PKEY_CTX_new_id(EVP_PKEY_DH, eng)`, even though the parameters you are eventually going to generate are `EVP_PKEY_DHX` parameters. But here we don't seem to allow that. It's very inconsistent.